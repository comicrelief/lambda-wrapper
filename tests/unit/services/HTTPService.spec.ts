import axios from 'axios';

import {
  COMICRELIEF_TEST_METADATA_HEADER,
  Context,
  DependencyInjection,
  HTTPService,
  RequestService,
} from '@/src';
import mockEvent from '@/tests/mocks/aws/event.json';

const mockContext = { invokedFunctionArn: 'my-function' } as Context;

const getService = (event = mockEvent, context = mockContext) => {
  const di = new DependencyInjection({
    dependencies: {
      HTTPService,
      RequestService,
    },
  }, event, context);
  return di.get(HTTPService);
};

describe('unit.services.HTTPService', () => {
  afterEach(() => jest.clearAllMocks());

  describe('request', () => {
    const testCases = {
      'GET request': { method: 'GET', url: '/' },
      'POST request': { method: 'POST', url: '/' },
      'PUT request': { method: 'PUT', url: '/' },
      'PATCH request': { method: 'PATCH', url: '/' },
      'HEAD request': { method: 'HEAD', url: '/' },
      'DELETE request': { method: 'DELETE', url: '/' },
      'with URL': { url: '/some/nested/path' },
      'with baseURL': { baseUrl: 'https://comicrelief.com/test', url: '/additional/url' },
      'overriding timeout': { timeout: 99 },
      'with headers': { headers: { Authorization: 'Bearer test' } },
      'with undefined headers': { headers: undefined },
    };

    Object.entries(testCases).forEach(([description, config]) => {
      it(description, async () => {
        const expected = { response: {} };
        const mock = jest.spyOn(axios, 'request').mockResolvedValue(expected);
        const service = getService();

        const response = await service.request(config);

        expect(response).toEqual(expected);
        expect(mock.mock.calls).toMatchSnapshot('config');
      });

      it(`adds the test header, ${description}`, async () => {
        const metadata = JSON.stringify({ user: 'Dante Alighieri' });
        const event = {
          ...mockEvent,
          headers: {
            ...mockEvent.headers,
            [COMICRELIEF_TEST_METADATA_HEADER]: metadata,
          },
        };
        const expected = { response: {} };
        const mock = jest.spyOn(axios, 'request').mockResolvedValue(expected);
        const service = getService(event);

        const response = await service.request(config);

        expect(response).toEqual(expected);
        expect(mock.mock.calls).toMatchSnapshot('config');
      });
    });
  });
});
