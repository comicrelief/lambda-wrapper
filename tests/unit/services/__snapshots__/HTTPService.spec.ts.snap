// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.services.HTTPService request DELETE request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "DELETE",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request GET request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "GET",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request HEAD request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "HEAD",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request PATCH request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "PATCH",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request POST request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "POST",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request PUT request: config 1`] = `
[
  [
    {
      "headers": {},
      "method": "PUT",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, DELETE request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "DELETE",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, GET request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "GET",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, HEAD request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "HEAD",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, PATCH request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "PATCH",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, POST request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "POST",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, PUT request: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "method": "PUT",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, overriding timeout: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "timeout": 99,
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, with URL: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "timeout": 10000,
      "url": "/some/nested/path",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, with baseURL: config 1`] = `
[
  [
    {
      "baseUrl": "https://comicrelief.com/test",
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "timeout": 10000,
      "url": "/additional/url",
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, with headers: config 1`] = `
[
  [
    {
      "headers": {
        "Authorization": "Bearer test",
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "timeout": 10000,
    },
  ],
]
`;

exports[`unit.services.HTTPService request adds the test header, with undefined headers: config 1`] = `
[
  [
    {
      "headers": {
        "x-comicrelief-test-metadata": "{"user":"Dante Alighieri"}",
      },
      "timeout": 10000,
    },
  ],
]
`;

exports[`unit.services.HTTPService request overriding timeout: config 1`] = `
[
  [
    {
      "headers": {},
      "timeout": 99,
    },
  ],
]
`;

exports[`unit.services.HTTPService request with URL: config 1`] = `
[
  [
    {
      "headers": {},
      "timeout": 10000,
      "url": "/some/nested/path",
    },
  ],
]
`;

exports[`unit.services.HTTPService request with baseURL: config 1`] = `
[
  [
    {
      "baseUrl": "https://comicrelief.com/test",
      "headers": {},
      "timeout": 10000,
      "url": "/additional/url",
    },
  ],
]
`;

exports[`unit.services.HTTPService request with headers: config 1`] = `
[
  [
    {
      "headers": {
        "Authorization": "Bearer test",
      },
      "timeout": 10000,
    },
  ],
]
`;

exports[`unit.services.HTTPService request with undefined headers: config 1`] = `
[
  [
    {
      "headers": undefined,
      "timeout": 10000,
    },
  ],
]
`;
